pkg/tenant/tenants.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/rotationalio/ensign/pkg/tenant/api/v1"
+	"github.com/rotationalio/ensign/pkg/tenant/db"
+	"github.com/rs/zerolog/log"
+)
+
+// CreateTenant creates a new tenant in the database
+func (s *Server) CreateTenant(c *gin.Context) {
+	var (
+		err    error
+		tenant *api.Tenant
+	)
+
+	if err = c.BindJSON(&tenant); err != nil {
+		log.Warn().Err(err).Msg("could not parse tenant create request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse request"))
+		return
+	}
+
+	if tenant.ID == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(api.ErrTenantIDRequired))
+	}
+
+	if tenant.Name == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("tenant name is required"))
+	}
+
+	if tenant.EnvironmentType == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("environment type is required"))
+	}
+
+	// Add Tenant to the database
+	if err = db.CreateTenant(c.Request.Context(), &db.Tenant{ID: ulid.Make()}); err != nil {
+		log.Error().Err(err).Msg("could not save tenant")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not add tenant"))
+		return
+	}
+
+	c.JSON(http.StatusCreated, tenant)
+}
+
+func (s *Server) TenantList(c *gin.Context) {
+
+	c.JSON(http.StatusNotImplemented, "not implemented yet")
+}
+
+func (s *Server) TenantCreate(c *gin.Context) {
+	var (
+		err    error
+		tenant *api.Tenant
+	)
+
+	// TODO: Add authentication and authorization middleware
+	if err = c.BindJSON(&tenant); err != nil {
+		log.Warn().Err(err).Msg("could not parse tenant create request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not bind request"))
+	}
+
+	if tenant.ID == "" {
+		log.Error().Err(err).Msg("tenant does not have an ID")
+	}
+
+	if tenant.Name == "" {
+		log.Warn().Err(err).Msg("tenant does not have an name")
+	}
+
+	if tenant.EnvironmentType == "" {
+		log.Error().Err(err).Msg("tenant does not have an environment type")
+	}
+
+	// TODO: Add tenant to the database
+	if err = db.CreateTenant(c.Request.Context(), &db.Tenant{ID: ulid.Make()}); err != nil {
+		log.Error().Err(err).Msg("could not save tenant")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not add tenant"))
+	}
+
+	c.JSON(http.StatusCreated, tenant)
+}
+
+func (s *Server) TenantDetail(c *gin.Context) {
+	c.JSON(http.StatusNotImplemented, "not implemented yet")
+}
+
+func (s *Server) TenantUpdate(c *gin.Context) {
+	c.JSON(http.StatusNotImplemented, "not implemented yet")
+}
+
+func (s *Server) TenantDelete(c *gin.Context) {
+	var (
+		err error
+		out *api.Reply
+	)
+
+	tenantID := c.Param("tenantID")
+
+	if tenantID == "" {
+		log.Error().Err(err).Msg(db.ErrMissingID.Error())
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not retrieve tenant"))
+		return
+	}
+
+	// TODO: Replace uuid with ulid
+	req := &db.Tenant{
+		ID: ulid.Make(),
+	}
+
+	// Delete Tenant and
+	if err = db.DeleteTenant(c.Request.Context(), req.ID); err != nil {
+		log.Error().Err(err).Msg(db.ErrMissingID.Error())
```

Comment:
We should probably talk about systematizing error logging -- I actually haven't figured this out yet, but it's probably important. For now what I'll say is to keep in mind that any `log.Error()` message will go to the slack channel and should cause us to take action. E.g. we're hoping that if there is a `Delete` error it means something has gone horribly wrong. We want to make sure we give as much information as possible in the log message:

```suggestion
		log.Error().Err(err).Str("tenantID", tenantID.String()).Msg("could not delete tenant")
```

Reply:
Thank you for providing more details about this.